core/src/main/scala/spire/optional/Perm.scala

```diff
@@ -0,0 +1,112 @@
+package spire
+package optional
+
+import scala.collection.compat._
+import spire.scalacompat.SeqLike
+
+import cats.kernel.Eq
+import spire.algebra.{Action, Group}
+import spire.algebra.partial.PartialAction
+import spire.syntax.cfor._
+import spire.util._
+
+/**
+ * Represents a permutation encoded as a map from preimages to images, including
+ * only pairs that are moved by the permutation (so the identity is Map.empty).
+ *
+ * Note that although the preimage and image contain only moved points, the
+ * domain and range of a Perm is all non-negative integers.
+ */
+class Perm private(private val mapping: Map[Int, Int]) extends (Int => Int) {
+  require(mapping.values.toSet == mapping.keySet, "Image and preimage must be the same.")
+  require(mapping.keys.forall(_ >= 0), "Perm indices must be non-negative.")
+
+  private lazy val inverseMapping = mapping.map(_.swap)
+
+  /**
+   * Apply this permutation to an `Int`.
+   *
+   * Note that that this acts as the identity for integers not in the image.
+   */
+  override def apply(k: Int): Int = mapping.getOrElse(k, k)
+
+  override def toString: String = {
+    mapping.toSeq.sorted
+      .map { case (k, v) => s"$k -> $v"}
+      .mkString("Perm(", ", ", ")")
+  }
+
+  /**
+   * The set of points moved by this permutation.
+   *
+   * Note that this is equivalent to the preimage.
+   */
+  def image: Set[Int] = mapping.keySet
+
+  /**
+   * The preimage of `k` (so that `apply(invert(k)) == k`).
+   */
+  def invert(k: Int): Int = inverseMapping.getOrElse(k, k)
+
+  /**
+   * The inverse permutation, which composes with this to yield the identity.
+   *
+   * Note that this uses a lazy inverse map, so the first call may be O(n)
+   * rather than O(1).
+   */
+  def inverse: Perm = new Perm(inverseMapping)
+
+  /**
+   * Compose this with another `Perm` with this permutation applied last.
+   */
+  def compose(that: Perm): Perm = Perm((this.image | that.image).map(k => k -> this(that(k))).toMap)
+
+  /**
+   * Compose this with another `Perm` with this permutation applied first.
+   */
+  def andThen(that: Perm): Perm = that.compose(this)
+
+  /**
+   * Permute a seq as long as all the moved points are valid indices into the seq.
+   */
+  def permute[A, SA <: SeqLike[A, SA]](seq: SA)(implicit cbf: Factory[A, SA]): Opt[SA] = {
+    if (this.image.isEmpty) return Opt(seq)
+    if (this.image.max >= seq.size) return Opt.empty[SA]
+    val builder = cbf.newBuilder
+    cforRange(0 until seq.size) { k =>
+      builder += seq(this(k))
+    }
+    Opt(builder.result())
+  }
+}
+
+object Perm {
+  def apply(mapping: Map[Int, Int]): Perm = {
+    new Perm(mapping.filter(Function.tupled(_ != _)))
+  }
+  def apply(pairs: (Int, Int)*): Perm = apply(Map(pairs: _*))
```

Comment:
My remaining qualm with this PR is that this constructor can start to become redundant.
We might rather use a constructor that takes advantage of disjoint cycle notation. eg., this perm:
```scala
Perm(
  1 -> 3,
  2 -> 4,
  3 -> 5,
  4 -> 1,
  5 -> 2
)
```
Could be more simply expressed as
```scala
Perm(1, 3, 5, 2, 4)
```

Reply:
We could define the latter with:
```scala
  def apply(points: Int*): Perm = {
    if (points.isEmpty) Perm(Map.empty[Int, Int])
    else Perm(points.zip(points.tail :+ points.head).toMap)
  }
```
But that has the disadvantage that it can only express permutations with a single cycle.
Dennis Rosset has a nice approach in alasc that uses apply to continue composing cycles.

Reply:
I've gone ahead and implemented this with two required Ints and a repeated Int parameter:

```scala
def apply(n0: Int, n1: Int, ns: Int*): Perm
```

This prevents some erroneous or redundant constructions with the extra benefit of avoiding ambiguity with apply overloads (the constructor using repeated pairs: `apply(pairs: (Int, Int)*)`, and the apply method on a single int `apply (n: Int): Int`). These ultimately defer to safer constructors that enforce properties, so this should be uncontroversial.